File: src/main/java/de/gesellix/docker/client/filesocket/FileSocketFactory.java
package de.gesellix.docker.client.filesocket;

import okhttp3.Dns;

import javax.net.SocketFactory;
import java.net.InetAddress;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.Collections;
import java.util.List;

abstract class FileSocketFactory extends SocketFactory implements Dns {

    @Override
    public List<InetAddress> lookup(String hostname) throws UnknownHostException {
        if (hostname.endsWith(FileSocket.SOCKET_MARKER)) {
            return Collections.singletonList(InetAddress.getByAddress(hostname, new byte[] {0, 0, 0, 0}));
        }
        else {
            return Dns.SYSTEM.lookup(hostname);
        }
    }

    @Override
    public Socket createSocket(String s, int i) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Socket createSocket(String s, int i, InetAddress inetAddress, int i1) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Socket createSocket(InetAddress inetAddress, int i) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Socket createSocket(InetAddress inetAddress, int i, InetAddress inetAddress1, int i1) {
        throw new UnsupportedOperationException();
    }
}

File: src/main/kotlin/de/gesellix/docker/client/filesocket/FileSocketFactory.kt
package de.gesellix.docker.client.filesocket

import okhttp3.Dns
import java.net.InetAddress
import java.net.Socket

import javax.net.SocketFactory

abstract class FileSocketFactory : SocketFactory(), Dns {

    @Override
    override fun lookup(hostname: String): List<InetAddress> {
        return if (hostname.endsWith(FileSocket.SOCKET_MARKER))
            listOf(InetAddress.getByAddress(hostname, kotlin.ByteArray(4)))
        else Dns.SYSTEM.lookup(hostname)
    }

    @Override
    override fun createSocket(s: String, i: Int): Socket {
        throw  UnsupportedOperationException()
    }

    @Override
    override fun createSocket(s: String, i: Int, inetAddress: InetAddress, i1: Int): Socket {
        throw UnsupportedOperationException()
    }

    @Override
    override fun createSocket(inetAddress: InetAddress, i: Int): Socket {
        throw UnsupportedOperationException()
    }

    @Override
    override fun createSocket(inetAddress: InetAddress, i: Int, inetAddress1: InetAddress, i1: Int): Socket {
        throw UnsupportedOperationException()
    }
}
